app/src/main/java/org/keynote/godtools/android/activity/MainActivity.java

```diff
@@ -245,10 +266,97 @@ private void loadInitialFragmentIfNeeded() {
+    void showNextFeatureDiscovery() {
+        if (!prefs().isFeatureDiscovered(FEATURE_LANGUAGE_SETTINGS) &&
+                canShowFeatureDiscovery(FEATURE_LANGUAGE_SETTINGS)) {
+            mFeatureDiscoveryHandler.postDelayed(() -> showFeatureDiscovery(FEATURE_LANGUAGE_SETTINGS, false), 10000);
+        }
+    }
+
+    /**
+     * Returns if the activity is in a state that it can actually show the specified feature discovery.
+     */
+    private boolean canShowFeatureDiscovery(@NonNull final String feature) {
+        switch (feature) {
+            case FEATURE_LANGUAGE_SETTINGS:
+                return mToolbar != null && (mDrawerLayout == null || !mDrawerLayout.isDrawerOpen(Gravity.START));
+        }
+
+        // assume we can show it if we don't have any specific rules about it
+        return true;
+    }
+
+    private void showFeatureDiscovery(@NonNull final String feature, final boolean force) {
+        // short-circuit if this activity is not started
+        if (!getLifecycle().getCurrentState().isAtLeast(RESUMED)) {
+            return;
+        }
+
+        // short-circuit if feature discovery is already visible
+        if (mFeatureDiscovery != null) {
+            return;
+        }
+
+        // short-circuit if this feature was discovered and we aren't forcing it
+        if (prefs().isFeatureDiscovered(feature) && !force) {
+            return;
+        }
+
+        // short-circuit if we can't show this feature discovery right now,
+        // and try to show the next feature discovery that can be shown.
+        if (!canShowFeatureDiscovery(feature)) {
+            showNextFeatureDiscovery();
+            return;
+        }
+
+        // dispatch specific feature discovery
+        switch (feature) {
+            case FEATURE_LANGUAGE_SETTINGS:
+                assert mToolbar != null : "canShowFeatureDiscovery() verifies mToolbar is not null";
+                if (mToolbar.findViewById(R.id.action_switch_language) != null) {
+                    final TapTarget target = TapTarget.forToolbarMenuItem(
+                            mToolbar, R.id.action_switch_language,
+                            getString(R.string.feature_discovery_title_language_settings),
+                            getString(R.string.feature_discovery_desc_language_settings));
+                    mFeatureDiscovery =
+                            TapTargetView.showFor(this, target, new LanguageSettingsFeatureDiscoveryListener());
+                    mFeatureDiscoveryActive = feature;
+                } else {
+                    // TODO: we currently don't (can't?) distinguish between when the menu item doesn't exist and when
+                    // TODO: the menu item just hasn't been drawn yet.
+
+                    // the toolbar action isn't available yet.
+                    // re-attempt this feature discovery on the next iteration of the main Looper.
+                    mToolbar.post(() -> showFeatureDiscovery(feature, force));
```

Comment:
Could this end up in an infinite loop type situation? I don't see any "sanity check" variable to make sure this doesn't get called more than 'n' times.

Reply:
it'll end up in an infinite loop as long as the activity is resumed and the language settings button never appears. But I believe in that scenario it will fire only once per frame.

Reply:
I was originally trying to find a callback to tie into that fired after the action button was available, but I couldn't find one.